doc/story008-repl-traces.md

```diff
@@ -0,0 +1,262 @@
+# Story008: Simulator Traces in REPL
+
+| Revision | Date       | Author                  |
+|---------:|:----------:|:------------------------|
+|        1 | 2023-05-05 | Igor Konnov             |
+
+This is a UX design document for the interaction with states and traces in REPL.
+This document helps us to understand the user stories before implementing them.
+
+Over the last months, we had several requests that could be unified under a
+common theme that is related to state and trace inspection.
+
+ - [#446][]
+ - [#795][]
+ - [#347][]
+ - [#303][]
+ - [#288][]
+
+This document is using [kettle][] as a running example. We assume that
+a Quint REPL session is run in the directory of the example:
+
+```sh
+$ cd ./tutorials/repl
+$ quint
+Quint REPL 0.9.1
+Type ".exit" to exit, or ".help" for more information
+>>> import kettle.* from "./kettle"
+
+```
+
+## Story 1: Inspect the current state
+
+*As a protocol reader, I want to print the protocol state without remembering
+the names of the state variables, so I can discover which state variables are
+around and which values they carry.*
+
+Assume that we have initialized the protocol state with `init`:
+
+```bluespec
+>>> init
+true
+```
+
+**Current interface**. Currently, we can inspect the state, by evaluating
+the variable names:
+
+```bluespec
+>>> temperature
+20
+>>> heatingOn
+false
+>>> beeping
+false
+```
+
+However, it is often more convenient to see the value of all the state variables
+at once, e.g., when we do not remember them. The issues [#446][] and [#795][]
+highlight this idea.
+
+A workaround is to introduce a definition like the following one:
+
+```bluespec
+// a handy definition that captures the state in a record
+val kettleState = {
+  heatingOn: heatingOn,
+  beeping: beeping,
+  temperature: temperature
+}
+```
+
+**Proposed interface**. The simplest yet powerful interface is to introduce the
+system state as a record that is accessible via a standard name. We decided to
+use the name `q::state`. Here is an example:
+
+```bluespec
+>>> q::state
+{
+  beeping: false,
```

Comment:
This is the first time I see something living in the `q::` toplevel.

Wouldn't it be more consistent to use a REPL command `.state` for this?
That has the added benefit that I can "discover" it through `.help`.

Reply:
+1 for .state and not risking colliding with any user defined namespaces.

+1 for it being a regular record.

how about an intermediate approach, using an empty namespace, that is, `::state`?

What if I have multiple modules imported? Will `q::state` be a record of records?

If `::state` is used for the top-level namespace, then it becomes almost intuitive to use `M::state` for a module imported as `import bla as M from file`